docs/design/proposals/split-repo/split-repo.md

```diff
@@ -0,0 +1,103 @@
+specific service controller's codebase.
+
+### Problem #2: Common runtime import inflexibility
+
+Furthermore, due to the code generator and ACK common runtime code being in the
+same repo as the service controllers that `import` this code, it's not easy to
+depend on a specific version of that common runtime from within the service
+controller code itself. Individual service controller package directories would
+need their own `go.mod` files and managing this would be a pain and wouldn't
+solve Problem #1 anyway.
+
+## Solution
+
+We propose splitting the monorepo into a set of source code repositories that
+can be separately Git-tagged and have their releases managed according to the
+preferences of the team that owns that component or individual ACK service
+controller.
+
+Specifically, we propose the following changes:
+
+1) All code for generating code will move into a new
+   `github.com/aws/aws-controllers-k8s-generate` source repository.
+
+   This includes all content in the existing `/cmd/ack-generate`, `/templates`,
+   `/pkg/generate` and `/pkg/model` directories of the existing
+   `github.com/aws/aws-controllers-k8s` source repository.
+
+2) The ACK common runtime and core ACK types will remain in the
+   `github.com/aws/aws-controllers-k8s` source repository.
+
+   The `/test` directory will contain only common testing framework code (both
+   Python and Bash-based for the immediate future).
+
+3) Each ACK service controller will move into a new
+   `github.com/aws/aws-controllers-k8s-$SERVICE` source repository.
+
+   There will be a single Git branch `main` for these individual service
+   controller source repositories. Teams that own individual service
+   controllers will be free to create additional Git branches for release
+   series -- e.g. `stable` or `preview` -- as they see fit.
+
+   Each separate service controller source repository shall have its own
+   `/test` directory containing tests for that service controller. These tests
+   will use the common testing frameworks from the main
+   `github.com/aws/aws-controllers-k8s` source repository. In the case of the
+   Python tests, this is as simple as actually consuming a versioned release of
+   the ACK common Python e2e testing framework. For the Bash-based tests, this
+   will likely mean some sort of file copying hack.
+
+4) The common build and image/chart publication scripts in
+   `github.com/aws/aws-controllers-k8s` will remain in that repo but will be
+   adapted as needed to allow pointing at different source repositories instead
+   of different subdirectories in the main monorepo. We will continue to
+   publish Docker images and Helm charts to the same single ECR registry and
+   continue to use the same basic image tagging scheme:
+   `$SERVICE-$RELEASE_VERSION`. The only difference is that `$RELEASE_VERSION`
+   will now be service controller-specific instead of a single release version
+   pertaining to all service controllers.
```

Comment:
For tagging to have binaries that corresponds to different release statuses (example: `Preview`, `Beta`, `GA`) should the tagging scheme have `$SERVICE-$RELEASE_STATUS-$RELEASE_VERSION` format?

Reply:
tagging on controller Docker images will not correspond to "Preview", "Beta" or "GA". Instead, there will be a SemVer-tagged images -- e.g. `controller/s3-v1.2.5`.

Helm Charts will get an **additional** Chart release tagged with a "-stable" suffix that will refer to the latest Git tag in the stable branch of the controller source repo. When we release new versions of a "stable" controller Helm chart, we will publish Charts that are tagged like "$SERVICE-stable-$IMAGE_SEMVER" in addition to a Chart that is aliased to "$SERVICE-stable".

For example, let's say that the latest Git tag on the S3 controller's `stable` branch is `v1.2.5`.

There will be a `chart/s3-stable-v1.2.5` Chart published to our ECR repo. The `image:` field in the `deployment.yaml` in the Chart would point to the `controller/s3-v1.2.5` image.

There will also be a `chart/s3-stable` Chart published to our ECR repo. The `image:` field in the `deployment.yaml` in this Chart would **also** point to the `controller/s3-v1.2.5` image.

Now assume that another SHA1 commit that is merged into the `stable` branch of the service controller repo is Git tagged with `v1.2.6`. We would publish a new `chart/s3-stable-v1.2.6` Chart published to our ECR repo. The `image:` field in the `deployment.yaml` in the Chart would point to the `controller/s3-v1.2.6` image.

We would then **update** the `chart/s3-stable` Chart to point to this latest `controller/s3-v1.2.6` image and change the `chart/s3-stable` Chart manifest tag.

We could likewise keep a `chart/s3-latest` Chart tag that points to an automatically-built controller image with a `vX.Y.Z-{COMMITSHA}`  image tag to allow for developers to install the latest controller code from the `main` branch.
 
Remember that releases are just tags --  both Git tags and tags on images or image manifests. Tags are cheap.

Reply:
Is the proposal to have just have two tags: `stable`, `latest` or is it to have tags corresponding to Dev preview, Beta, GA?

Reply:
The proposal is to have just a `$SERVICE-stable` Chart tag and `$SERVICE-latest` Chart tag. Note: no stable image tags or latest image tags.